source/common/upstream/load_balancer_impl.cc

```diff
@@ -890,6 +892,140 @@ HostConstSharedPtr RandomLoadBalancer::peekOrChoose(LoadBalancerContext* context
   return hosts_to_use[random_hash % hosts_to_use.size()];
 }
 
+ShuffleShardLoadBalancer::ShuffleShardLoadBalancer(
+    LoadBalancerType lb_type, const PrioritySet& priority_set,
+    const PrioritySet* local_priority_set, ClusterStats& stats, Runtime::Loader& runtime,
+    Random::RandomGenerator& random,
+    const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config,
+    const envoy::config::cluster::v3::Cluster::LbShuffleShardConfig& config)
+    : ZoneAwareLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
+                                common_config),
+      lb_type_(lb_type),
+      endpoints_per_cell_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, endpoints_per_cell, 2)),
+      use_zone_as_dimension_(config.use_zone_as_dimension()),
+      use_dimensions_(config.dimensions().size()),
+      shuffle_sharder_(ShuffleSharder<HostConstSharedPtr>(12345)) {
+
+  // ENVOY_LOG(debug, "endpoints_per_cell: {}", endpoints_per_cell_);
+  // ENVOY_LOG(debug, "use_zone_as_dimension: {}", use_zone_as_dimension_);
+
+  for (auto dimension : config.dimensions())
+    dimensions_.push_back(dimension);
+  if (use_zone_as_dimension_)
+    dimensions_.push_back("_envoy_zone");
+  if (!dimensions_.size())
+    dimensions_.push_back("_envoy_unit_coord");
+
+  lattice_ = new Lattice<HostConstSharedPtr>(dimensions_);
```

Comment:
this naked pointer should be replaced with a smart pointer. Maybe provider a helper function to build the dimensions and then you can constructor initialize lattice_